File: packages/plugin-sdk-python/tests/test_db_value.py
"""Golden vectors for typed DbValue (Cap'n wire, i64 bounds, UTF-8, unknown unions)."""

from __future__ import annotations

import asyncio
import unittest

from bookclerk_plugin_sdk.db_value import (
    DatabaseBinding,
    RetryToken,
    canonical_execute_request_hash,
    decode_db_value,
    decode_execute_result_reply,
    decode_execute_request,
    encode_db_value,
    encode_execute_result_reply,
    encode_execute_request,
    parse_db_value,
)

I64_MIN = -(2**63)
I64_MAX = 2**63 - 1

GOLDEN_I64_MIN = (
    "00000000040000000000000002000100000002000000000000000000000000800000000000000000"
)
GOLDEN_I64_MAX = (
    "000000000400000000000000020001000000020000000000ffffffffffffff7f0000000000000000"
)
GOLDEN_TEXT_B64 = (
    "0000000006000000000000000200010000000400000000000000000000000000010000004a000000"
    "6236343a414141410000000000000000"
)
GOLDEN_BYTES_012 = (
    "0000000005000000000000000200010000000500000000000000000000000000010000001a000000"
    "0001020000000000"
)
GOLDEN_BOOL_TRUE = (
    "00000000040000000000000002000100010001000000000000000000000000000000000000000000"
)
GOLDEN_NULL_BYTES = (
    "00000000040000000000000002000100050000000000000000000000000000000000000000000000"
)
GOLDEN_EXECUTE_REQUEST = (
    "000000001d0000000000000004000300000000000000000000000000000000000000000000000000"
    "0000000000000000090000001a0000000900000022000000090000001f0000006f70000000000000"
    "616263000000000004000000010002000200020001000000050000004a000000090000004f000000"
    "53454c454354203f00000000000000000c0000000200010000000200000000000000000000000080"
    "0000000000000000000004000000000000000000000000000d000000720000000000050000000000"
    "0000000000000000090000000a0000006236343a6e6f742d6279746573000000ff00000000000000"
)


class DbValueGoldens(unittest.TestCase):
    def test_typed_null_bytes(self) -> None:
        self.assertEqual(
            parse_db_value({"kind": "null", "value": "bytes"}),
            {"kind": "null", "value": "bytes"},
        )

    def test_i64_min_max(self) -> None:
        for n in (I64_MIN, -1, 0, 1, I64_MAX):
            self.assertEqual(
                parse_db_value({"kind": "int64", "value": n}),
                {"kind": "int64", "value": n},
            )

    def test_utf8_and_embedded_nul(self) -> None:
        text = parse_db_value({"kind": "text", "value": "héllo\x00world"})
        self.assertEqual(text["value"], "héllo\x00world")

    def test_embedded_zero_bytes(self) -> None:
        blob = parse_db_value({"kind": "bytes", "value": "b64:AAEC"})
        self.assertEqual(blob["kind"], "bytes")
        self.assertEqual(blob["value"], b"\x00\x01\x02")
        raw = parse_db_value({"kind": "bytes", "value": b"\x00\x01\x02"})
        self.assertEqual(raw["value"], b"\x00\x01\x02")

    def test_unknown_union_member(self) -> None:
        with self.assertRaisesRegex(ValueError, "unknown DbValue union member"):
            parse_db_value({"kind": "xml", "value": "<a/>"})

    def test_capnp_goldens(self) -> None:
        self.assertEqual(encode_db_value({"kind": "int64", "value": I64_MIN}).hex(), GOLDEN_I64_MIN)
        self.assertEqual(encode_db_value({"kind": "int64", "value": I64_MAX}).hex(), GOLDEN_I64_MAX)
        self.assertEqual(
            encode_db_value({"kind": "text", "value": "b64:AAAA"}).hex(), GOLDEN_TEXT_B64
        )
        self.assertEqual(
            encode_db_value({"kind": "bytes", "value": b"\x00\x01\x02"}).hex(),
            GOLDEN_BYTES_012,
        )
        self.assertEqual(encode_db_value({"kind": "boolean", "value": True}).hex(), GOLDEN_BOOL_TRUE)
        self.assertEqual(
            encode_db_value({"kind": "null", "value": "bytes"}).hex(), GOLDEN_NULL_BYTES
        )
        self.assertNotEqual(
            encode_db_value({"kind": "text", "value": "b64:AAAA"}),
            encode_db_value({"kind": "bytes", "value": b"\x00\x01\x02"}),
        )
        self.assertEqual(decode_db_value(encode_db_value({"kind": "int64", "value": I64_MIN}))["value"], I64_MIN)
        self.assertEqual(
            decode_db_value(encode_db_value({"kind": "text", "value": "b64:AAAA"}))["value"],
            "b64:AAAA",
        )
        self.assertEqual(
            decode_db_value(encode_db_value({"kind": "bytes", "value": b"\x00\x01\x02"}))["value"],
            b"\x00\x01\x02",
        )

    def test_execute_request_golden(self) -> None:
        request = {
            "operationId": "op",
            "requestHash": "abc",
            "statements": [
                {
                    "sql": "SELECT ?",
                    "parameters": [
                        {"kind": "int64", "value": I64_MIN},
                        {"kind": "text", "value": "b64:not-bytes"},
                        {"kind": "bytes", "value": b"\xff"},
                    ],
                    "kind": "select",
                    "maxRows": 1,
                    "resultSelection": "rows",
                }
            ],
            "deadlineUnixMs": 0,
        }
        encoded = encode_execute_request(request)
        self.assertEqual(encoded.hex(), GOLDEN_EXECUTE_REQUEST)
        back = decode_execute_request(encoded)
        self.assertEqual(back["operationId"], "op")
        self.assertEqual(back["statements"][0]["parameters"][0]["value"], I64_MIN)
        self.assertEqual(back["statements"][0]["parameters"][1]["value"], "b64:not-bytes")
        self.assertEqual(back["statements"][0]["parameters"][2]["value"], b"\xff")

        seen = {"n": 0}

        async def transport(req):
            seen["n"] = len(encode_execute_request(req))
            return {
                "operationId": req["operationId"],
                "statements": [],
                "timing": {
                    "attemptElapsedUs": 0,
                    "dbExecutionUs": 0,
                    "dbTimingSource": "test",
                },
            }

        binding = DatabaseBinding(transport, operation_id="op", request_hash="abc")

        async def run() -> None:
            reply = await binding.execute(request["statements"])
            self.assertEqual(reply["operationId"], "op")
            self.assertEqual(seen["n"], len(encoded))

        asyncio.run(run())

    def test_two_sequential_batches_use_distinct_operation_ids(self) -> None:
        seen: list[str] = []
        hashes: list[str] = []

        async def transport(req):
            seen.append(req["operationId"])
            self.assertNotEqual(req["operationId"], "op")
            return {
                "operationId": req["operationId"],
                "statements": [
                    {"rows": [], "columns": [], "rowsAffected": 0}
                ],
                "timing": {
                    "attemptElapsedUs": 0,
                    "dbExecutionUs": 0,
                    "dbTimingSource": "test",
                },
            }

        async def recording(req):
            hashes.append(req["requestHash"])
            return await transport(req)

        async def run() -> None:
            binding = DatabaseBinding(recording)
            first = [{"kind": "int64", "value": 1}]
            second = [{"kind": "int64", "value": 2}]
            await binding.prepare("INSERT INTO t VALUES (?)").bind(*first).run()
            await binding.prepare("INSERT INTO t VALUES (?)").bind(*second).run()
            self.assertEqual(len(seen), 2)
            self.assertNotEqual(seen[0], seen[1])
            self.assertNotEqual(seen[0], "op")
            self.assertNotEqual(seen[1], "op")
            self.assertEqual(hashes, ["", ""])
            retry = RetryToken(seen[0], canonical_execute_request_hash({
                "operationId": seen[0],
                "requestHash": "",
                "statements": [
                    {
                        "sql": "INSERT INTO t VALUES (?)",
                        "parameters": first,
                        "kind": "execute",
                        "maxRows": 0,
                        "resultSelection": "affectedRows",
                    }
                ],
                "deadlineUnixMs": 0,
            }))
            replayed = []

            async def replay_transport(req):
                replayed.append((req["operationId"], req["requestHash"]))
                return await transport(req)

            replay = DatabaseBinding(replay_transport)
            await replay.prepare("INSERT INTO t VALUES (?)").bind(*first).run(retry=retry)
            self.assertEqual(replayed[0][0], seen[0])
            self.assertEqual(replayed[0][1], retry.request_hash)

        asyncio.run(run())

    def test_capnp_reader_rejects_truncated_multisegment_and_oversized_count(self) -> None:
        with self.assertRaisesRegex(ValueError, "truncated"):
            decode_db_value(b"\x00\x00")
        multi = (
            (1).to_bytes(4, "little")
            + (1).to_bytes(4, "little")
            + (1).to_bytes(4, "little")
            + (0).to_bytes(4, "little")
            + b"\x00" * 16
        )
        with self.assertRaisesRegex(ValueError, "multi-segment"):
            decode_db_value(multi)
        # Composite list tag with count 2^29 and a 1-word payload (D=1).
        import struct

        def pack_word(n: int) -> bytes:
            return struct.pack("<Q", n)

        # single-segment header, size 4 words: root ptr + far-looking list + tag + pad
        # Build: nseg=1, size0=4. Root is a composite list pointer with huge count.
        list_ptr = 1 | (0 << 2) | (7 << 32) | (1 << 35)  # offset 0, composite, D=1
        tag = (0x20000000 << 2) | (1 << 32) | (0 << 48)  # count ~2^29, 1 data word
        buf = bytearray()
        buf += struct.pack("<II", 0, 4)
        buf += pack_word(list_ptr)  # word 0 root — not a struct; decode_db_value expects struct
        buf += pack_word(0)
        buf += pack_word(tag)
        buf += pack_word(0)
        with self.assertRaises(ValueError):
            decode_execute_request(bytes(buf))
        # Backward (negative) struct offset that leaves the segment.
        back_ptr = 0 | ((0x3FFFFFFF) << 2) | (2 << 32) | (1 << 48)
        back = bytearray()
        back += struct.pack("<II", 0, 2)
        back += pack_word(back_ptr)
        back += pack_word(0)
        with self.assertRaisesRegex(ValueError, "out of segment|far pointer|expected struct"):
            decode_db_value(bytes(back))

    def test_canonical_hash_ignores_deadline_golden(self) -> None:
        req = {
            "operationId": "op",
            "requestHash": "abc",
            "statements": [
                {
                    "sql": "SELECT 1",
                    "parameters": [],
                    "kind": "select",
                    "maxRows": 1,
                    "resultSelection": "rows",
                }
            ],
            "deadlineUnixMs": 0,
        }
        golden = "e368ef90b76963c5e93c5e6db37fdb6d7f809d23c10295352a0ba3cd26885f02"
        self.assertEqual(canonical_execute_request_hash(req), golden)
        req["deadlineUnixMs"] = 9_999_999_999
        self.assertEqual(canonical_execute_request_hash(req), golden)

    def test_first_sends_max_rows_one_and_mixed_batch_keeps_per_statement_intent(self) -> None:
        seen: list[dict] = []

        async def transport(req):
            seen.append(req["statements"][0] if len(req["statements"]) == 1 else req["statements"])
            n = len(req["statements"])
            return {
                "operationId": req["operationId"],
                "statements": [
                    {
                        "rows": [{"values": [{"kind": "int64", "value": 1}]}],
                        "columns": [{"name": "n", "dbType": "int64"}],
                        "rowsAffected": 0,
                                            }
                    for _ in range(n)
                ],
                "timing": {
                    "attemptElapsedUs": 0,
                    "dbExecutionUs": 0,
                    "dbTimingSource": "test",
                },
            }

        async def run() -> None:
            binding = DatabaseBinding(transport)
            row = await binding.prepare("SELECT n FROM t").first()
            self.assertEqual(seen[0]["maxRows"], 1)
            self.assertEqual(seen[0]["resultSelection"], "rows")
            self.assertEqual(row["n"]["value"], 1)

            seen.clear()
            await binding.batch(
                [
                    binding.prepare("INSERT INTO t VALUES (?)").bind(
                        {"kind": "int64", "value": 1}
                    ).as_run(),
                    binding.prepare("SELECT n FROM t").as_all(),
                ]
            )
            stmts = seen[0]
            self.assertEqual(stmts[0]["resultSelection"], "affectedRows")
            self.assertEqual(stmts[0]["maxRows"], 0)
            self.assertEqual(stmts[1]["resultSelection"], "rows")

        asyncio.run(run())

    def test_execute_result_reply_preserves_i64_bytes_and_error_code(self) -> None:
        reply = {
            "operationId": "op",
            "statements": [
                {
                    "rows": [
                        {
                            "values": [
                                {"kind": "int64", "value": I64_MIN},
                                {"kind": "int64", "value": I64_MAX},
                                {"kind": "bytes", "value": b"\xff"},
                                {"kind": "text", "value": "b64:not-bytes"},
                                {"kind": "null", "value": "int64"},
                            ]
                        }
                    ],
                    "columns": [
                        {"name": "lo", "dbType": "int64"},
                        {"name": "hi", "dbType": "int64"},
                        {"name": "blob", "dbType": "bytes"},
                        {"name": "txt", "dbType": "text"},
                        {"name": "n", "dbType": "int64"},
                    ],
                    "rowsAffected": 0,
                                    }
            ],
            "timing": {
                "attemptElapsedUs": 1,
                "dbExecutionUs": 2,
                "dbTimingSource": "test",
            },
        }
        back = decode_execute_result_reply(encode_execute_result_reply(reply))
        cells = back["statements"][0]["rows"][0]["values"]
        self.assertEqual(cells[0]["value"], I64_MIN)
        self.assertEqual(cells[1]["value"], I64_MAX)
        self.assertEqual(cells[2]["value"], b"\xff")
        self.assertEqual(cells[3]["value"], "b64:not-bytes")
        from bookclerk_plugin_sdk.workerd import PluginError

        with self.assertRaises(PluginError) as ctx:
            decode_execute_result_reply(
                encode_execute_result_reply(("unavailable", "retry me"))
            )
        self.assertEqual(ctx.exception.code, "unavailable")
        self.assertEqual(ctx.exception.wire_code, "unavailable")


if __name__ == "__main__":
    unittest.main()

File: packages/plugin-sdk/scripts/test-db-value.mjs
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const root = dirname(fileURLToPath(import.meta.url));
const dist = join(root, "../dist");

const {
  parseDbValue,
  encodeDbValue,
  decodeDbValue,
} = await import(join(dist, "db-value.js"));
const {
  encodeExecuteRequest,
  decodeExecuteRequest,
  createDatabaseBinding,
  canonicalExecuteRequestHash,
  encodeExecuteResultReply,
  decodeExecuteResultReply,
} = await import(join(dist, "db-execute.js"));

function hex(bytes) {
  return Buffer.from(bytes).toString("hex");
}

const I64_MIN = -0x8000_0000_0000_0000n;
const I64_MAX = 0x7fff_ffff_ffff_ffffn;

const GOLDEN = {
  i64Min:
    "00000000040000000000000002000100000002000000000000000000000000800000000000000000",
  i64Max:
    "000000000400000000000000020001000000020000000000ffffffffffffff7f0000000000000000",
  textB64:
    "0000000006000000000000000200010000000400000000000000000000000000010000004a0000006236343a414141410000000000000000",
  bytes012:
    "0000000005000000000000000200010000000500000000000000000000000000010000001a0000000001020000000000",
  boolTrue:
    "00000000040000000000000002000100010001000000000000000000000000000000000000000000",
  nullBytes:
    "00000000040000000000000002000100050000000000000000000000000000000000000000000000",
  executeRequest:
    "000000001d00000000000000040003000000000000000000000000000000000000000000000000000000000000000000090000001a0000000900000022000000090000001f0000006f70000000000000616263000000000004000000010002000100020001000000050000004a000000090000004f00000053454c454354203f00000000000000000c00000002000100000002000000000000000000000000800000000000000000000004000000000000000000000000000d0000007200000000000500000000000000000000000000090000000a0000006236343a6e6f742d6279746573000000ff00000000000000",
};

const nullBytes = parseDbValue({ kind: "null", value: "bytes" });
assert.deepEqual(nullBytes, { kind: "null", value: "bytes" });

for (const n of [-1, 0, 1]) {
  assert.deepEqual(parseDbValue({ kind: "int64", value: n }), {
    kind: "int64",
    value: BigInt(n),
  });
}

assert.deepEqual(parseDbValue({ kind: "int64", value: I64_MIN }), {
  kind: "int64",
  value: I64_MIN,
});
assert.deepEqual(parseDbValue({ kind: "int64", value: I64_MAX }), {
  kind: "int64",
  value: I64_MAX,
});

const text = parseDbValue({ kind: "text", value: "héllo\u0000world" });
assert.equal(text.kind, "text");
assert.equal(text.value, "héllo\u0000world");

const blob = parseDbValue({ kind: "bytes", value: "b64:AAEC" });
assert.equal(blob.kind, "bytes");
assert.deepEqual(Array.from(blob.value), [0, 1, 2]);

assert.throws(
  () => parseDbValue({ kind: "xml", value: "<a/>" }),
  /unknown DbValue union member/,
);

assert.equal(hex(encodeDbValue({ kind: "int64", value: I64_MIN })), GOLDEN.i64Min);
assert.equal(hex(encodeDbValue({ kind: "int64", value: I64_MAX })), GOLDEN.i64Max);
assert.equal(
  hex(encodeDbValue({ kind: "text", value: "b64:AAAA" })),
  GOLDEN.textB64,
);
assert.equal(
  hex(encodeDbValue({ kind: "bytes", value: Uint8Array.of(0, 1, 2) })),
  GOLDEN.bytes012,
);
assert.equal(hex(encodeDbValue({ kind: "boolean", value: true })), GOLDEN.boolTrue);
assert.equal(
  hex(encodeDbValue({ kind: "null", value: "bytes" })),
  GOLDEN.nullBytes,
);

assert.notEqual(
  hex(encodeDbValue({ kind: "text", value: "b64:AAAA" })),
  hex(encodeDbValue({ kind: "bytes", value: Uint8Array.of(0, 1, 2) })),
);

assert.equal(decodeDbValue(encodeDbValue({ kind: "int64", value: I64_MIN })).value, I64_MIN);
assert.equal(decodeDbValue(encodeDbValue({ kind: "int64", value: I64_MAX })).value, I64_MAX);
assert.equal(decodeDbValue(encodeDbValue({ kind: "text", value: "b64:AAAA" })).value, "b64:AAAA");
assert.deepEqual(
  Array.from(decodeDbValue(encodeDbValue({ kind: "bytes", value: Uint8Array.of(0, 1, 2) })).value),
  [0, 1, 2],
);

const request = {
  operationId: "op",
  requestHash: "abc",
  statements: [
    {
      sql: "SELECT ?",
      parameters: [
        { kind: "int64", value: I64_MIN },
        { kind: "text", value: "b64:not-bytes" },
        { kind: "bytes", value: Uint8Array.of(0xff) },
      ],
      kind: "select",
      maxRows: 1,
      resultSelection: "rows",
    },
  ],
  deadlineUnixMs: 0,
};

const reqBytes = encodeExecuteRequest(request);
assert.equal(hex(reqBytes), GOLDEN.executeRequest);
const back = decodeExecuteRequest(reqBytes);
assert.equal(back.operationId, "op");
assert.equal(back.statements[0].parameters[0].value, I64_MIN);
assert.equal(back.statements[0].parameters[1].value, "b64:not-bytes");
assert.deepEqual(Array.from(back.statements[0].parameters[2].value), [0xff]);

let encodedByBinding = 0;
const binding = createDatabaseBinding(
  {
    async execute(req) {
      encodedByBinding = encodeExecuteRequest(req).byteLength;
      return {
        operationId: req.operationId,
        statements: [],
        timing: { attemptElapsedUs: 0, dbExecutionUs: 0, dbTimingSource: "test" },
      };
    },
  },
  { operationId: "op", requestHash: "abc" },
);
const reply = await binding.execute(request.statements);
assert.equal(reply.operationId, "op");
assert.equal(encodedByBinding, reqBytes.byteLength);

const sequentialIds = [];
const seqBinding = createDatabaseBinding({
  async execute(req) {
    sequentialIds.push(req.operationId);
    assert.notEqual(req.operationId, "op");
    assert.equal(req.requestHash, "");
    return {
      operationId: req.operationId,
      statements: [{ rows: [], columns: [], rowsAffected: 0 }],
      timing: { attemptElapsedUs: 0, dbExecutionUs: 0, dbTimingSource: "test" },
    };
  },
});
await seqBinding.prepare("INSERT INTO t VALUES (?)").bind({ kind: "int64", value: 1n }).run();
await seqBinding.prepare("INSERT INTO t VALUES (?)").bind({ kind: "int64", value: 2n }).run();
assert.equal(sequentialIds.length, 2);
assert.notEqual(sequentialIds[0], sequentialIds[1]);

const GOLDEN_DEADLINE_HASH =
  "648cd28b3223c825c55ea99a7c6e52321ea733656f5965abfe4c7ed4ca21d111";
const deadlineReq = {
  operationId: "op",
  requestHash: "abc",
  statements: [
    {
      sql: "SELECT 1",
      parameters: [],
      kind: "select",
      maxRows: 1,
      resultSelection: "rows",
    },
  ],
  deadlineUnixMs: 0,
};
assert.equal(await canonicalExecuteRequestHash(deadlineReq), GOLDEN_DEADLINE_HASH);
deadlineReq.deadlineUnixMs = 9999999999;
assert.equal(await canonicalExecuteRequestHash(deadlineReq), GOLDEN_DEADLINE_HASH);

const firstSeen = [];
const firstBinding = createDatabaseBinding({
  async execute(req) {
    firstSeen.push({
      maxRows: req.statements[0].maxRows,
      resultSelection: req.statements[0].resultSelection,
    });
    return {
      operationId: req.operationId,
      statements: [
        {
          rows: [
            {
              values: [{ kind: "int64", value: 1n }],
            },
          ],
          columns: [{ name: "n", dbType: "int64" }],
          rowsAffected: 0,
          cursor: "",
        },
      ],
      timing: { attemptElapsedUs: 0, dbExecutionUs: 0, dbTimingSource: "test" },
    };
  },
});
const row = await firstBinding.prepare("SELECT n FROM t").first();
assert.equal(firstSeen[0].maxRows, 1);
assert.equal(firstSeen[0].resultSelection, "rows");
assert.equal(row.n.value, 1n);

const col = await firstBinding.prepare("SELECT n FROM t").first("n");
assert.equal(col.value, 1n);

const allResult = await firstBinding.prepare("SELECT n FROM t").all();
assert.equal(allResult.success, true);
assert.ok(Array.isArray(allResult.results));
assert.equal(allResult.results[0].n.value, 1n);
assert.equal(allResult.meta.rows_read, 1);

const runSeen = [];
const runBinding = createDatabaseBinding({
  async execute(req) {
    runSeen.push({
      resultSelection: req.statements[0].resultSelection,
      kind: req.statements[0].kind,
    });
    return {
      operationId: req.operationId,
      statements: [
        {
          rows: [
            {
              values: [{ kind: "int64", value: 2n }],
            },
          ],
          columns: [{ name: "n", dbType: "int64" }],
          rowsAffected: 0,
          cursor: "",
        },
      ],
      timing: { attemptElapsedUs: 0, dbExecutionUs: 0, dbTimingSource: "test" },
    };
  },
});
const runResult = await runBinding.prepare("SELECT n FROM t").run();
assert.equal(runSeen[0].resultSelection, "rows");
assert.equal(runSeen[0].kind, "select");
assert.equal(runResult.results[0].n.value, 2n);

const mixedSeen = [];
const mixBinding = createDatabaseBinding({
  async execute(req) {
    mixedSeen.push(
      req.statements.map((s) => ({
        selection: s.resultSelection,
        maxRows: s.maxRows,
      })),
    );
    return {
      operationId: req.operationId,
      statements: req.statements.map(() => ({
        rows: [],
        columns: [],
        rowsAffected: 1,
        cursor: "",
      })),
      timing: { attemptElapsedUs: 0, dbExecutionUs: 0, dbTimingSource: "test" },
    };
  },
});
await mixBinding.batch([
  mixBinding.prepare("INSERT INTO t VALUES (?)").bind({ kind: "int64", value: 1n }),
  mixBinding.prepare("SELECT n FROM t"),
]);
assert.deepEqual(mixedSeen[0], [
  { selection: "rows", maxRows: 0 },
  { selection: "rows", maxRows: 0 },
]);

await mixBinding.batch([
  mixBinding.prepare("INSERT INTO t VALUES (?)").bind({ kind: "int64", value: 1n }).asRun(),
  mixBinding.prepare("SELECT n FROM t").asAll(),
]);
assert.deepEqual(mixedSeen[1], [
  { selection: "affectedRows", maxRows: 0 },
  { selection: "rows", maxRows: 0 },
]);

const batchResults = await mixBinding.batch([
  mixBinding.prepare("INSERT INTO t VALUES (?)").bind({ kind: "int64", value: 1n }),
  mixBinding.prepare("SELECT n FROM t"),
]);
assert.equal(batchResults.length, 2);
assert.equal(batchResults[0].success, true);
assert.equal(batchResults[0].meta.changes, 1);
assert.equal(batchResults[1].success, true);

const i64Reply = {
  operationId: "op",
  statements: [
    {
      rows: [
        {
          values: [
            { kind: "int64", value: I64_MIN },
            { kind: "int64", value: I64_MAX },
            { kind: "bytes", value: Uint8Array.of(0xff) },
            { kind: "text", value: "b64:not-bytes" },
            { kind: "null", value: "int64" },
          ],
        },
      ],
      columns: [
        { name: "lo", dbType: "int64" },
        { name: "hi", dbType: "int64" },
        { name: "blob", dbType: "bytes" },
        { name: "txt", dbType: "text" },
        { name: "n", dbType: "int64" },
      ],
      rowsAffected: 0,
      cursor: "",
    },
  ],
  timing: { attemptElapsedUs: 1, dbExecutionUs: 2, dbTimingSource: "test" },
};
const atomicOk = encodeExecuteResultReply({ ok: i64Reply });
const decodedOk = decodeExecuteResultReply(atomicOk);
assert.equal(decodedOk.statements[0].rows[0].values[0].value, I64_MIN);
assert.equal(decodedOk.statements[0].rows[0].values[1].value, I64_MAX);
assert.deepEqual(
  Array.from(decodedOk.statements[0].rows[0].values[2].value),
  [0xff],
);
assert.equal(decodedOk.statements[0].rows[0].values[3].value, "b64:not-bytes");
assert.equal(decodedOk.statements[0].rows[0].values[4].value, "int64");

const atomicErr = encodeExecuteResultReply({
  err: { code: "unavailable", message: "retry me" },
});
try {
  decodeExecuteResultReply(atomicErr);
  assert.fail("expected PluginError");
} catch (err) {
  assert.equal(err.code, "unavailable");
  assert.equal(err.wireCode, "unavailable");
  assert.equal(err.message, "retry me");
}

assert.throws(() => decodeDbValue(new Uint8Array([0, 0])), /truncated/);
const multi = new Uint8Array(24);
new DataView(multi.buffer).setUint32(0, 1, true);
assert.throws(() => decodeDbValue(multi), /multi-segment/);

assert.equal(typeof readFileSync, "function");
console.log("db-value goldens ok");
